src/distance.cpp

```diff
@@ -0,0 +1,620 @@
+       handle_t nextHandle = *nextNodes[0];
+       size_t j = index(start, make_pair(graph->get_id(nextHandle), 
+                                         graph->get_is_reverse(nextHandle)));
+
+       int64_t firstDist = distances->at(j);//Length of first node 
+       return totalDist - firstDist;
+    }
+}
+
+ChainDistances::ChainDistances(unordered_map<id_t, size_t>* s, 
+                                                          vector<int64_t>* p) {
+    
+    snarl_to_index = s;
+    prefix_sum = p;
+}
+
+int64_t ChainDistances::chainDistance(pair<id_t, bool> start, 
+                                                      pair<id_t, bool> end) {
+    //Distance between start of start and start of end in a chain
+    size_t i1 = snarl_to_index->at(start.first);
+    size_t i2 = snarl_to_index->at(end.first); 
+    if ((start.second && end.second) || (!start.second && !end.second)) {
+        //If start and end are facing the same direction
+        int64_t dNoRev = abs(prefix_sum->at(i1) - prefix_sum->at(i2)); 
+        return dNoRev;
+    } else {
+        //TODO Currently assumes no reversing edges in the chain
+        return -1;
+    }
+}
+
+int64_t ChainDistances::chainDistance_short(VG* graph, pair<id_t, bool> start,
+                                                      pair<id_t, bool> end) {
+    //Distance between end of start node to beginning of end node in chain
+    
+    size_t i1 = snarl_to_index->at(start.first);
+    size_t i2 = snarl_to_index->at(end.first); 
+    int64_t d1 = chainDistance(start, end);
+    int64_t d2 = chainDistance(end, start);
+    if (d1 == -1 && d2 == -1) {
+        return -1;
+    } else if (d1 == -1) {
+        return d2 - graph->get_node(i1)->sequence().size();
+    } else if (d2 == -1) {
+        return d1 - graph->get_node(i2)->sequence().size();
+    } else {
+        return min(d2 - graph->get_node(i1)->sequence().size(), 
+                   d1 - graph->get_node(i2)->sequence().size());
+    }
+}
+
+int64_t ChainDistances::chainLength() {
+    //Get the length of a chain including length of last node
+    return prefix_sum->back();
+}
+
+
+int64_t calculateIndex(DistanceIndex* di, VG* graph, SnarlManager* sm, 
+         const Chain* chain) {
+    /*Add chain to distance index and recursively add all distances in 
+      snarls contained within chain
+    */
+    auto cmp = [] (pair<const handle_t*,int64_t> x,
+                                           pair<const handle_t*,int64_t> y) {
+        //Comparison function for the priority of a pair of handle, distance
+        return (x.second < y.second); 
+    };
+
+    vector<int64_t> chain_prefix_sum (1,0); //initialize to [0]
+
+    unordered_map<id_t, size_t> chain_snarl_to_index;
+    for (int i = 0; i < chain->size(); i++) { 
+        //for each snarl in the chain TODO ChainIterator?
+        const Snarl* snarl = chain->at(i);
+        chain_snarl_to_index.insert( make_pair<id_t, size_t>
```

Comment:
There's really nothing *wrong* with `insert`, but I think it's often cleaner to read
```
 chain_snarl_to_index[snarl->start().node_id()] = chain_snarl_to_index.size();
```
Also, when you have a template function like `make_pair`, you rarely need to specify it's templates. It's also kind of easy to mess things up by specifying templates incorrectly. As a rule, I usually recommend trying to compile without templates on STL functions, and then only add them in if the compilation fails. The compiler's type inference is pretty smart. It usually only fails if you're using an implicit conversion between types (i.e. like using a 32-bit int as the argument to a function that takes 64-bit ints)